File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedActionBar.java
package net.easecation.ghosty.recording.level.updated;

import cn.nukkit.Player;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.recording.level.LevelRecordNode;

public class LevelUpdatedActionBar implements LevelUpdated {

    private String message;
    int fadeInTime;
    int stayTime;
    int fadeOutTime;

    private LevelUpdatedActionBar(String message, int fadeInTime, int stayTime, int fadeOutTime) {
        this.message = message;
        this.fadeInTime = fadeInTime;
        this.stayTime = stayTime;
        this.fadeOutTime = fadeOutTime;
    }

    public LevelUpdatedActionBar(BinaryStream stream) {
        this.read(stream);
    }

    public static LevelUpdatedActionBar of(String message, int fadeInTime, int stayTime, int fadeOutTime) {
        return new LevelUpdatedActionBar(message, fadeInTime, stayTime, fadeOutTime);
    }

    @Override
    public int getUpdateTypeId() {
        return LevelUpdated.TYPE_ACTION_BAR;
    }

    @Override
    public void processTo(LevelRecordNode node) {
        node.offerLevelGlobalCallback((level) -> {
            for (Player player : level.getPlayers().values()) {
                player.sendActionBar(this.message, this.fadeInTime, this.stayTime, this.fadeOutTime);
            }
        });
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putString(message);
        stream.putLInt(fadeInTime);
        stream.putLInt(stayTime);
        stream.putLInt(fadeOutTime);
    }

    @Override
    public void read(BinaryStream stream) {
        this.message = stream.getString();
        this.fadeInTime = stream.getLInt();
        this.stayTime = stream.getLInt();
        this.fadeOutTime = stream.getLInt();
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof LevelUpdatedActionBar o)) return false;
        return this.message.equals(o.message) && this.fadeInTime == o.fadeInTime && this.stayTime == o.stayTime && this.fadeOutTime == o.fadeOutTime;
    }

    @Override
    public String toString() {
        return "LevelUpdatedActionBar{" +
            "message='" + message + '\'' +
            ", fadeInTime=" + fadeInTime +
            ", stayTime=" + stayTime +
            ", fadeOutTime=" + fadeOutTime +
            '}';
    }

    @Override
    public int hashCode() {
        return this.message.hashCode() + this.fadeInTime + this.stayTime + this.fadeOutTime;
    }
}

File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedActionBar.kt
package net.easecation.ghosty.recording.level.updated

import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.recording.level.LevelRecordNode

@Serializable
data class LevelUpdatedActionBar(
    var message: String,
    var fadeInTime: Int,
    var stayTime: Int,
    var fadeOutTime: Int,
) : LevelUpdated {
    override fun getUpdateTypeId(): Int = LevelUpdated.TYPE_ACTION_BAR

    override fun processTo(node: LevelRecordNode) {
        node.offerLevelGlobalCallback { level ->
            level.players.values.forEach { player ->
                player.sendActionBar(message, fadeInTime, stayTime, fadeOutTime)
            }
        }
    }

    override fun backwardTo(node: LevelRecordNode) = Unit

    companion object {
        @JvmStatic
        fun of(message: String, fadeInTime: Int, stayTime: Int, fadeOutTime: Int) =
            LevelUpdatedActionBar(message, fadeInTime, stayTime, fadeOutTime)

        @JvmStatic
        val ADAPTER: LevelUpdateAdapter<LevelUpdatedActionBar> = Adapter
    }

    private object Adapter : LevelUpdateAdapter<LevelUpdatedActionBar> {
        override fun write(updated: LevelUpdatedActionBar, stream: BinaryStream) {
            stream.putString(updated.message)
            stream.putLInt(updated.fadeInTime)
            stream.putLInt(updated.stayTime)
            stream.putLInt(updated.fadeOutTime)
        }

        override fun read(stream: BinaryStream, formatVersion: Int) = LevelUpdatedActionBar(
            message = stream.string,
            fadeInTime = stream.lInt,
            stayTime = stream.lInt,
            fadeOutTime = stream.lInt,
        )
    }
}
